src/ring_theory/dedekind_domain/adic_valuation.lean

```diff
@@ -222,4 +239,226 @@ begin
+infer_instance
+
+lemma v_valued_K_topological_add_group :
+  @topological_add_group K v.v_valued_K_topological_space _ := infer_instance
+
+/-- The uniform space structure on `K` corresponding to the `v`-adic valuation. -/
+def v_valued_K_uniform_space : uniform_space K :=
+@topological_add_group.to_uniform_space K _ v.v_valued_K_topological_space
+  v.v_valued_K_topological_add_group
+
+lemma v_valued_K_uniform_add_group : @uniform_add_group K v.v_valued_K_uniform_space _ :=
+@topological_add_group_is_uniform K _ v.v_valued_K_topological_space
+  v.v_valued_K_topological_add_group
+
+lemma v_valued_K_completable_top_field : @completable_top_field K _ v.v_valued_K_uniform_space :=
+@valued.completable K _ _ _ (v_valued_K v)
+
+lemma v_valued_K_separated_space : @separated_space K v.v_valued_K_uniform_space :=
+@valued_ring.separated K _ _ _ (v_valued_K v)
+
+variables (K)
+/-- The completion of `K` with respect to its `v`-adic valuation. -/
+def adic_completion := @uniform_space.completion K v.v_valued_K_uniform_space
+
+instance : field (v.adic_completion K) :=
+@field_completion K _ v.v_valued_K_uniform_space v.v_valued_K_topological_division_ring _
+  v.v_valued_K_uniform_add_group
+
+instance : inhabited (v.adic_completion K) := ⟨0⟩
+
+variables {K}
+instance valued_adic_completion : valued (v.adic_completion K) (with_zero (multiplicative ℤ)) :=
+⟨@valued.extension_valuation K _ _ _ (v_valued_K v)⟩
+
+lemma valued_adic_completion_def {x : v.adic_completion K} :
+  valued.v (x) = @valued.extension K _ _ _ (v_valued_K v)  x := rfl
+
+instance adic_completion_topological_space : topological_space (v.adic_completion K) :=
+@valued.topological_space (v.adic_completion K) _ _ _ v.valued_adic_completion
+
+instance adic_completion_topological_division_ring :
+  @topological_division_ring (v.adic_completion K) _ v.adic_completion_topological_space :=
+@valued.topological_division_ring (v.adic_completion K) _ _ _ v.valued_adic_completion
+
+instance adic_completion_topological_ring :
+  @topological_ring (v.adic_completion K) v.adic_completion_topological_space _ :=
+v.adic_completion_topological_division_ring.to_topological_ring
+
+instance adic_completion_topological_add_group :
+  @topological_add_group (v.adic_completion K) v.adic_completion_topological_space _ :=
+@topological_ring.to_topological_add_group (v.adic_completion K) _
+  v.adic_completion_topological_space v.adic_completion_topological_ring
+
+instance adic_completion_uniform_space : uniform_space (v.adic_completion K) :=
+@topological_add_group.to_uniform_space (v.adic_completion K) _
+  v.adic_completion_topological_space v.adic_completion_topological_add_group
+
+instance adic_completion_uniform_add_group :
+  @uniform_add_group (v.adic_completion K) v.adic_completion_uniform_space _ :=
+@topological_add_group_is_uniform (v.adic_completion K) _ v.adic_completion_topological_space
+  v.adic_completion_topological_add_group
+
```

Comment:
Does the completion automatically get a `complete_space` (or whatever it is called) instance?

Reply:
Just glancing at this now I think there might be some diamonds at play.

I would have expected that `adic_completion_uniform_space` to be defined as:
```lean
instance adic_completion_uniform_space : uniform_space (v.adic_completion K) :=
@uniform_space.completion.uniform_space K v.v_valued_K_uniform_space
```
This does permit:
```lean
example : complete_space (v.adic_completion K) :=
@uniform_space.completion.complete_space K v.v_valued_K_uniform_space -- not sure why `by apply_instance` fails
```
but it looks like this causes `adic_completion_uniform_add_group` to choke on a diamond. Solving these diamonds is exactly the job of `uniform_add_group` so I'm optimistic it won't be too hard to resolve.

I'll look into this tomorrow (assuming it's still pending).

Reply:
Thanks! Yes, I tried to change the definition of `adic_completion_uniform_space` to `@uniform_space.completion.uniform_space K v.v_valued_K_uniform_space` earlier, but since that broke `adic_completion_uniform_add_group` I changed it back (because I'm going to use this instance). I hope there's a way to make the `complete_space` instance work as well, we should definitely have it.

Reply:
Ah, it seems that everything works if I do
```
instance adic_completion_uniform_add_group :
  @uniform_add_group (v.adic_completion K) v.adic_completion_uniform_space _ :=
@uniform_space.completion.uniform_add_group K v.v_valued_K_uniform_space _ 
  v.v_valued_K_uniform_add_group
```

Reply:
I have the impression you define too many instances. For example since `v.adic_completion K` already has a field instance, `inhabited (v.adic_completion K)` should be automatic, right? Similarly for `topological_division_ring` that impies `topological_ring` and `topological_add_group`. I would try to define just the strongest ones, and see if `infer_instance` find the others. In case it doesn't work you can very well define them by hand, by you should add an example to check you are not creating a diamond (see for example [here](https://github.com/leanprover-community/mathlib/blob/e0f78ab30b500480e68710c42dfdc5ec6dce1951/src/analysis/normed/group/quotient.lean#L323)).

Reply:
The `topological_ring` and `topological_add_group` instances are found by `infer_instance`, so I've removed them. However, `inhabited` is not (I added it in the first place because the linter complained. I thought it was surprising that this was not inferred from the `field` instance).

Reply:
Wow
```
example {K : Type*} [field K] : inhabited K := infer_instance
```
doesn't work!

Reply:
We probably just don't care about these instances...

Reply:
`adic_completion_uniform_space` should follow from `adic_completion_uniform_add_group`, that should in turn follows from (I think) `adic_completion_topological_division_ring`, that implies also `adic_completion_topological_space`. If it is not automatic we really have to make sure the two topological space structures (given by `adic_completion_topological_division_ring` and the uniform space structure) are definitionally equal.

Reply:
Yeah I think we don't care.

I note that `classical.inhabited_of_nonempty'` is correctly not an instance (consider the problems we would otherwise have given `has_zero.nonempty` and `has_one.nonempty`). For a field we could declare `0` or `1` as the default value but really `nonempty` is the right instance here rather than `inhabited`.

See also this old thread [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/inhabited.20and.20nonempty.20instances)

Reply:
It seems that if I infer `uniform_add_group` from `topological_division_ring` I don't get the correct `uniform_space` structure: in the following code, `adic_completion_complete_space` yields an error
```
instance adic_completion_topological_division_ring :
  topological_division_ring (v.adic_completion K) :=
@valued.topological_division_ring (v.adic_completion K) _ _ _ v.valued_adic_completion

instance adic_completion_uniform_add_group : uniform_add_group (v.adic_completion K) :=
infer_instance

instance adic_completion_uniform_space : uniform_space (v.adic_completion K) :=
infer_instance

instance adic_completion_complete_space : complete_space (v.adic_completion K) :=
@uniform_space.completion.complete_space K v.adic_valued_uniform_space
```

Reply:
Overriding an existing instance (i.e., a declaration that works with `infer_instance`) is just covering up a diamond so there is something wrong here. I'll look into this one this afternoon.

Reply:
I believe the following is the diamond:
```lean
import topology.algebra.valued_field

noncomputable theory

variables (K Γ₀ : Type*) [field K] [linear_ordered_comm_group_with_zero Γ₀] [hv : valued K Γ₀]
variables [completable_top_field K] [topological_division_ring K] [uniform_add_group K]

instance foo : valued (@uniform_space.completion K $ @valued.uniform_space _ _ _ _ hv) Γ₀ :=
⟨valued.extension_valuation⟩

example :
  @valued.uniform_space (@uniform_space.completion K $ @valued.uniform_space _ _ _ _ hv) _ _ _ _ =
  uniform_space.completion.uniform_space K :=
rfl -- Fails!
```
I'm out of time for now but I'll have a think about what to do later this evening.

Reply:
OK, thanks!